tools/deployment_tools.js

```diff
@@ -0,0 +1,499 @@
+     *                 Defaults to a IncrementingAddressGenerator.
+     *
+     * @returns {object} The "accounts" section for a parity chainspec.
+     */
+    toParityGenesisAccounts(options = {}) {
+        const defaultOptions = {
+            addressGenerator: new IncrementingAddressGenerator(),
+        };
+        const mergedOptions = Object.assign(defaultOptions, options);
+
+        // prepare contracts by ordering and generating addresses
+        const { addressGenerator } = mergedOptions;
+        this._orderContracts();
+        this.contracts.forEach(contract => contract.setAddress(addressGenerator));
+        this.contracts.forEach(contract => contract.instantiate());
+
+        const output = {
+        };
+
+        this.contracts
+            .forEach((contract) => {
+                const address = contract.getAddress();
+                const constructor = contract.constructorData;
+
+                output[address] = {
+                    balance: '0',
+                    constructor,
+                };
+            });
+
+        return output;
+    }
+
+    /**
+     * Generate transaction objects that can be passed as an argument
+     * to `web3.eth.sendTransaction()`.
+     * This allows for deployment on a live network.
+     *
+     * @param {string} fromAddress See {@ IncrementingNonceAddressGenerator#constructor}.
+     * @param {number} startingNonce See {@ IncrementingNonceAddressGenerator#constructor}.
+     *
+     * @returns {Array.<object>} The transaction objects that can be used for deployment.
+     */
+    toLiveTransactionObjects(fromAddress, startingNonce) {
+        const addressGenerator = new IncrementingNonceAddressGenerator(fromAddress, startingNonce);
+
+        this._orderContracts();
+        this.contracts.forEach(contract => contract.setAddress(addressGenerator));
+        this.contracts.forEach(contract => contract.instantiate());
+
+        const transactionObjects = [];
+        this.contracts
+            .forEach((contract, i) => {
+                const transactionObject = {
+                    // fields for web3.eth.sendTransaction()
+                    from: fromAddress,
+                    data: contract.constructorData,
+                    nonce: startingNonce + i,
+                };
+                const deploymentObject = {
+                    transactionObject,
+                    // metadata
+                    address: contract.getAddress(),
+                    contractName: contract.contractName,
+                };
+                transactionObjects.push(deploymentObject);
+            });
+
+        return transactionObjects;
+    }
+
+    /**
+     * Orders all contracts in the order of sequential deployment.
+     * This is determined by the amount of transitive dependencies (including self).
+     */
+    _orderContracts() {
+        this.contracts.sort((a, b) => a._getDependenciesCount() - b._getDependenciesCount());
```

Comment:
I don't understand how the dependencies count guarantees the order? Can't a contract with `n` dependencies depend on a contract with `n+1` or `n-1` dependencies? 😕

Reply:
It's not just the count of dependencies but transitive dependencies. There is some graph theory behind it that @pgev can probably better explain 😅

Reply:
Ahhhh, yes. Now it makes sense 👍 I forgot to take the transitivity into account 😅